File: apps/ios/Janark/UI/Components/FeedCard.swift
import SwiftUI

struct FeedCard: View {
    @Environment(\.janarkColors) private var c

    let post: FeedPostDto
    var onOpen: () -> Void
    var onAuthor: ((String) -> Void)? = nil
    var footer: AnyView? = nil

    var body: some View {
        let tags = post.tags
        let kind = resolveFeedKind(type: post.type, title: post.title, href: post.href, tags: tags)
        let bar: Color = {
            switch kind.kind {
            case .petition: return c.success
            case .vote: return c.fact
            case .report: return c.danger
            case .issue: return c.amber
            case .notice: return c.amberBright
            case .share: return c.news
            case .discussion, .other: return c.muted
            }
        }()
        let place = placeLabel(city: post.city, district: post.district, state: post.state, country: post.country)
        let time = relativeTime(post.createdAt)
        let metric = metricLabel(kind.metric, n: post.votes)

        SurfaceCard(onTap: onOpen) {
            HStack(alignment: .top, spacing: 12) {
                AccentBar(color: bar)
                VStack(alignment: .leading, spacing: 8) {
                    HStack(spacing: 8) {
                        KindPill(label: kind.label, color: bar)
                        if post.hot == true {
                            KindPill(label: "Hot", color: c.amber)
                        }
                        if let publicId = post.publicId {
                            Text(publicId)
                                .font(.system(size: 11, design: .monospaced))
                                .foregroundStyle(c.muted)
                        }
                    }
                    Text(post.title)
                        .font(.system(.title3, design: .serif).weight(.semibold))
                        .foregroundStyle(c.foreground)
                        .lineLimit(3)
                    if let excerpt = post.excerpt, !excerpt.isEmpty {
                        Text(excerpt)
                            .font(.subheadline)
                            .foregroundStyle(c.muted)
                            .lineLimit(3)
                    }
                    CivicMedia(url: post.mediaUrl, mediaType: post.mediaType)
                    HStack(spacing: 10) {
                        Text("\(metric.0) \(metric.1)")
                            .font(.caption)
                            .foregroundStyle(c.muted)
                        if let place {
                            Text(place)
                                .font(.caption)
                                .foregroundStyle(c.muted)
                        }
                        if let time {
                            Text(time)
                                .font(.caption)
                                .foregroundStyle(c.muted)
                        }
                    }
                    if let authorAnonId = post.authorAnonId {
                        Button(authorAnonId) { onAuthor?(authorAnonId) }
                            .font(.caption)
                            .foregroundStyle(c.amber)
                            .buttonStyle(.plain)
                    } else if let author = post.author, !author.isEmpty {
                        Text(author)
                            .font(.caption)
                            .foregroundStyle(c.amber)
                    }
                    if !tags.isEmpty {
                        Text(tags.prefix(5).map { $0.hasPrefix("#") ? $0 : "#\($0)" }.joined(separator: " "))
                            .font(.caption)
                            .foregroundStyle(c.navyMid)
                    }
                    if let footer {
                        footer
                    }
                }
            }
        }
    }
}

extension FeedCard {
    init(
        post: FeedPostDto,
        onOpen: @escaping () -> Void,
        onAuthor: ((String) -> Void)? = nil,
        @ViewBuilder footer: () -> some View
    ) {
        self.post = post
        self.onOpen = onOpen
        self.onAuthor = onAuthor
        self.footer = AnyView(footer())
    }
}

File: apps/android/app/src/main/java/org/janark/app/ui/components/FeedCard.kt
package org.janark.app.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.janark.app.core.placeLabel
import org.janark.app.core.relativeTime
import org.janark.app.data.net.FeedPostDto
import org.janark.app.domain.FeedKind
import org.janark.app.domain.metricLabel
import org.janark.app.domain.resolveFeedKind
import org.janark.app.ui.theme.JanarkTheme
import org.janark.app.ui.theme.toC

@Composable
fun FeedCard(
    post: FeedPostDto,
    onOpen: () -> Unit,
    onAuthor: ((String) -> Unit)? = null,
    footer: @Composable (() -> Unit)? = null,
) {
    val c = JanarkTheme.colors
    val kind = resolveFeedKind(post.type, post.title, post.href, post.tags)
    val bar = when (kind.kind) {
        FeedKind.Petition -> c.success
        FeedKind.Vote -> c.fact
        FeedKind.Report -> c.danger
        FeedKind.Issue -> c.amber
        FeedKind.Notice -> c.amberBright
        FeedKind.Share -> c.news
        FeedKind.Discussion, FeedKind.Other -> c.muted
    }
    val place = placeLabel(post.city, post.district, post.state, post.country)
    val time = relativeTime(post.createdAt)
    val metric = metricLabel(kind.metric, post.votes ?: 0)

    SurfaceCard(onClick = onOpen) {
        Row(verticalAlignment = Alignment.Top) {
            AccentBar(bar)
            Spacer(Modifier.width(12.dp))
            Column(Modifier.weight(1f)) {
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    KindPill(kind.label, bar)
                    if (post.hot == true) KindPill("Hot", c.amber)
                    post.publicId?.let {
                        Text(it, fontFamily = FontFamily.Monospace, fontSize = 11.sp, color = c.muted.toC())
                    }
                }
                Spacer(Modifier.height(8.dp))
                Text(
                    post.title,
                    color = c.foreground.toC(),
                    fontFamily = FontFamily.Serif,
                    fontWeight = FontWeight.SemiBold,
                    fontSize = 18.sp,
                    maxLines = 3,
                    overflow = TextOverflow.Ellipsis,
                )
                val excerpt = post.excerpt?.takeIf { it.isNotBlank() }
                if (excerpt != null) {
                    Spacer(Modifier.height(6.dp))
                    Text(excerpt, color = c.muted.toC(), maxLines = 3, overflow = TextOverflow.Ellipsis, fontSize = 14.sp)
                }
                CivicMedia(post.mediaUrl, post.mediaType, Modifier.padding(top = 10.dp))
                Spacer(Modifier.height(8.dp))
                Row(horizontalArrangement = Arrangement.spacedBy(10.dp), verticalAlignment = Alignment.CenterVertically) {
                    Text("${metric.first} ${metric.second}", color = c.muted.toC(), fontSize = 12.sp)
                    if (place != null) Text(place, color = c.muted.toC(), fontSize = 12.sp)
                    if (time != null) Text(time, color = c.muted.toC(), fontSize = 12.sp)
                }
                val author = post.authorAnonId ?: post.author
                if (!author.isNullOrBlank()) {
                    Text(
                        author,
                        color = c.amber.toC(),
                        fontSize = 12.sp,
                        modifier = Modifier.padding(top = 4.dp).then(
                            if (post.authorAnonId != null && onAuthor != null)
                                Modifier.clickableNoRipple { onAuthor(post.authorAnonId) }
                            else Modifier,
                        ),
                    )
                }
                if (!post.tags.isNullOrEmpty()) {
                    Text(
                        post.tags.take(5).joinToString(" ") { if (it.startsWith("#")) it else "#$it" },
                        color = c.navyMid.toC(),
                        fontSize = 12.sp,
                        modifier = Modifier.padding(top = 6.dp),
                    )
                }
                footer?.invoke()
            }
        }
    }
}

private fun Modifier.clickableNoRipple(onClick: () -> Unit) =
    androidx.compose.foundation.clickable(onClick = onClick)
